### PR TITLE
Re-use bundles at startup for podman runtime

### PR DESCRIPTION
**Pull request by windsource**
Issue-Id: #260

# Definition of Done

The PR shall be merged only if all items mentioned in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been followed. In case an item is not applicable as described, please provide a short explanation in the description.

- [x] All steps in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been handled

### CODE REVIEW

**Comment by windsource** [comment]
All done. Ready for review.

**Comment by windsource** [comment]
Currently this PR only handles agent restart and reuses containers that have the same configuration and are in state succeeded. But he agent might also restart workloads with `restartPolicy: ALWAYS` which currently does not reuse the existing container, even when workloads exits with state succeeded. To be consistent we also should implement reusing containers for that use-case.

**Comment by krucod3** [comment]
Should we create a ticket for the restarts? We can then add the constrains in (e.g., only for successfully finished workloads).

**Comment by windsource** [comment]
@krucod3 At first I thought about adding this right in this PR but we can also review it like it is today and have another isssue + PR for the restart policy.

**Review by krucod3 — COMMENTED** [review]
Review done, findings documented.

File: tests/resources/configs/device_restart_with_dependencies_modified.yaml
[the remark this answers is not in the document]
Comment (windsource) [reply tests/resources/configs/device_restart_with_dependencies_modified.yaml:1]: Agree to change the file name to `device_restart_with_dependencies_modified_web_service_init.yaml`.

File: justfile [hunk 1]
```diff
@@ -35,12 +38,26 @@ check-test-images:
 check-copyright-headers:
```
[the remark this answers is not in the document]
Comment (windsource) [reply justfile:38]: OK.

File: tests/resources/ankaios.resource [hunk 2]
```diff
@@ -477,6 +482,50 @@ the container of workload "${workload_name}" shall have a different id but same
...
+    WHILE    True
+        ${current_secs}=    Get Time Secs
+        ${elapsed_secs}=    Evaluate    ${current_secs} - ${start_time}
+        IF    ${elapsed_secs} >= 20    BREAK
+        ${current_workload_id}    ${current_ankaios_instance_name}    Get Container Id And Name By Workload Name From Podman    ${workload_name}
+        IF    '${current_workload_id}' == '' or '${current_ankaios_instance_name}' == ''
+            CONTINUE
+        END
+
+        ${workload_id}=    Get From Dictionary    ${ANKAIOS_INSTANCE_NAME_TO_PODMAN_ID_MAPPING}    ${current_ankaios_instance_name}
```
[the remark this answers is not in the document]
Comment (windsource) [reply tests/resources/ankaios.resource:519]: OK

File: tests/resources/ankaios.resource [hunk 3]
```diff
@@ -477,6 +482,50 @@ the container of workload "${workload_name}" shall have a different id but same
 
     Should Be True    ${id_changed}    msg=Workload '${workload_name}' has the same id '${workload_id}' and the same configuration on podman!
 
+the container of workload "${workload_name}" shall have a different id and different configuration on the podman runtime
```
[the remark this answers is not in the document]
Comment (windsource) [reply tests/resources/ankaios.resource:485]: OK. I will remove the part `and different configuration` from the name.

File: tests/resources/ankaios.resource [hunk 4]
```diff
@@ -477,6 +482,50 @@ the container of workload "${workload_name}" shall have a different id but same
...
+        IF    '${current_workload_id}' == '' or '${current_ankaios_instance_name}' == ''
+            CONTINUE
+        END
+
+        ${workload_id}=    Get From Dictionary    ${ANKAIOS_INSTANCE_NAME_TO_PODMAN_ID_MAPPING}    ${current_ankaios_instance_name}
+
+        IF    '${current_workload_id}' != '${workload_id}'
+            ${id_changed}=    Set Variable    ${True}
+            BREAK
+        END
```
[the remark this answers is not in the document]
Comment (windsource) [reply tests/resources/ankaios.resource:524]: I will fix that.

File: agent/src/runtime_manager.rs [hunk 5]
```diff
@@ -254,15 +265,31 @@ impl RuntimeManager {
...
+                                // [impl->swdd~agent-existing-workloads-replace-updated~3]
 
                                 log::info!(
                                     "Replacing existing workload '{}'.",
                                     workload_state.instance_name.workload_name()
                                 );
 
-                                /* Temporary workaround until direct start of bundles is implemented to prevent
+                                /* Workaround to prevent
                                 workload states from being overwritten by the delete. The decoupled create and a potential enqueue
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_manager.rs:293]: OK

File: agent/src/runtime_connectors/runtime_facade.rs [hunk 6]
```diff
@@ -40,17 +43,35 @@ use crate::workload::WorkloadCommandSender;
...
+        workload_state: WorkloadState,
+        workload_id: Option<String>,
+    ) -> ReusableWorkloadState {
+        ReusableWorkloadState {
+            workload_state,
+            workload_id,
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_connectors/runtime_facade.rs:63]: OK

File: agent/src/runtime_manager.rs [hunk 7]
```diff
@@ -310,21 +338,39 @@ impl RuntimeManager {
                 .eq(new_instance_name)
     }
 
+    fn is_reusable_workload(
+        workload_state_existing_workload: &WorkloadState,
+        workload_id_existing_workload: &Option<String>,
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_manager.rs:343]: Good point. Will change that.

File: agent/src/runtime_manager.rs [hunk 8]
```diff
@@ -155,7 +159,12 @@ impl RuntimeManager {
             deleted_workloads.len()
         );
 
-        self.execute_workloads(added_workloads, deleted_workloads, workload_state_db)
+        let new_added_workloads = added_workloads
+            .into_iter()
+            .map(|w| ReusableWorkloadSpec::new(w, None))
+            .collect();
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_manager.rs:165]: I think I cannot implement the existing `From` or `Into` traits for `Vec` as the implementation needs to reside either where the trait or the object resides and as both are system definitions that won't work. But I can use an own trait.

File: agent/src/runtime_connectors/podman/podman_runtime.rs [hunk 9]
```diff
@@ -95,15 +105,18 @@ impl PodmanRuntime {
...
-                    execution_state,
-                }),
+            let workload_id = &self.get_workload_id(instance_name).await?.id;
+            match PodmanCli::list_states_by_id(workload_id).await {
+                Ok(Some(execution_state)) => workload_states.push(ReusableWorkloadState::new(
+                    WorkloadState {
+                        instance_name: instance_name.clone(),
+                        execution_state,
+                    },
+                    Some(workload_id.to_string()),
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_connectors/podman/podman_runtime.rs:118]: OK.

File: agent/src/runtime_connectors/podman_cli.rs [hunk 10]
```diff
@@ -1075,6 +1106,52 @@ mod tests {
         assert_eq!(res, Ok("test_id".to_string()));
     }
 
+    // [utest->swdd~podman-create-workload-starts-existing-workload~1]
+    #[tokio::test]
+    async fn utest_start_container_success() {
+        let _guard = MOCKALL_CONTEXT_SYNC.get_lock_async().await;
+
+        static ID: &str = "test_id";
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_connectors/podman_cli.rs:1114]: OK. I will replace it with `const`.

File: agent/src/runtime_connectors/podman_cli.rs [hunk 11]
```diff
@@ -1075,6 +1106,52 @@ mod tests {
...
+                .expect_args(&["start", ID])
+                .exec_returns(Err(SAMPLE_ERROR_MESSAGE.into())),
+        );
+
+        let start_config = super::PodmanStartConfig {
+            general_options: vec![],
+            container_id: ID.into(),
+        };
+        let res = PodmanCli::podman_start(start_config, "test_workload_name").await;
+        assert!(matches!(res, Err(msg) if msg == SAMPLE_ERROR_MESSAGE));
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_connectors/podman_cli.rs:1152]: OK

File: agent/src/runtime_connectors/podman_kube/podman_kube_runtime.rs [hunk 12]
```diff
@@ -65,20 +66,32 @@ impl Display for PodmanKubeWorkloadId {
     }
 }
 
+impl FromStr for PodmanKubeWorkloadId {
+    type Err = String;
+    fn from_str(_s: &str) -> Result<Self, Self::Err> {
+        // Not implemented as we cannot restart workloads
+        Err("Not implemented for PodmanKubeWorkloadId".to_string())
```
[the remark this answers is not in the document]
Comment (windsource) [reply agent/src/runtime_connectors/podman_kube/podman_kube_runtime.rs:73]: OK

**Comment by windsource** [comment]
Fixed review findings.

File: agent/src/runtime_manager.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (krucod3) [reply agent/src/runtime_manager.rs]: It is just required that the type of the vector is in the same file as Vec is part of the std and is special.

**Review by krucod3 — APPROVED** [review]
LGTM 👍

**Pull request merged**